### PR TITLE
Beacon integration -> Adding and removing VCF variants from case page

### LINKED ISSUE #2185: Submit variants to beacon directly
- Beacon will be --> https://github.com/Clinical-Genomics/cgbeacon2
- Create a role for users so only allowed users can send data to the beacon
- Create a button so when user clicks on submit they'll submit a link to the VCF file and the list of all genes to filter by.
- Only variants from affected individuals from a case will be collected and inserted into the beacon database.
- Create option to remove all variants from a case

### PR DESCRIPTION
**Pull request by northwestwitch**
This PR adds the functionality to allow users (with `beacon_submitter` role) to submit variants to the Beacon. fix #2185

Users case decide to:
- submit variants for either only affected individuals or all individuals 
- submit variants from one or more of these categories: `SNVs`, `research SNVs`. If case is a cancer cases the available choices are: `cancer SNVs`, `research cancer SNVs`
- submit variants from one or more gene panels (gene panels used to run the case only) or the entire VCF (if they don't filter by any panel)
- Remove variants from one or all individuals from the beacon

image

**How to test --> locally because the beacon is yet to be installed on the new server farm**:
You'll need to install and run the beacon from a terminal distinct from the one used to run scout

**Installing the beacon software and running the server**
1. From a conda env with python 3.6> 
```
git clone https://github.com/Clinical-Genomics/cgbeacon2.git
cd cgbeacon2
pip install -e .
```
1. Update available genes in the beacon by running the command:
```
beacon update genes
```
1. Create a new dataset corresponding to the scout institute. Let's name it "cust000_GRCh37" (We'll need a new one named cust000_GRCh38 when we switch build. This is because each beacon dataset can only contain one type of genome build)
```
beacon add dataset -id cust000_GRCh37 -name "Test database with genome build GRCh37" -build GRCh37 -authlevel public -desc "It's a test"
```
1. Create a new user for the beacon api named scout: 
```
beacon add user -id scout -name "Scout browser" -url https://scout-stage.scilifelab.se
```
Note that at the moment that the user is created, a token is saved to the beacon database for this user:
image

1. Launch the app on localhost using a distinct port from the default used by scout (5000):
```
beacon run -p 6000
```

**From Scout terminal, with demo data loaded in database**
1. Grant test user the privilege to save variants to the beacon. Run this cli command:
```
 scout --demo update user -u clark.kent@mail.com -r beacon_submitter
```
1. Edit the demo config file by adding these lines:
```
BEACON_URL = "http://localhost:6000/apiv1.0"
BEACON_TOKEN = token_created_for_the_beacon_user
```
1. Modify the vcf_files.vcf_snv and vcf_files.vcf_snv_research values of demo case in database so that they include the full path of these files in your disk, see example from my instance:
image

1. Launch scout demo and go to demo case. You should see a new option at the bottom of the sidebar:
image

1. Try to share case variants to the beacon in different scenarios (from one or case individuals with or without gene panel filtering).

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [x] tests executed by mikaell

### CODE REVIEW

**Comment by northwestwitch** [comment]
This PR is working now. I (just 😢) need to improve the test instructions including the use of the auth token, write something in the docs and write the tests

**Comment by northwestwitch** [comment]
Yes! Coverage should be fine now. Ready for review!

**Comment by mikaell** [comment]
`pip install -r requirements.txt` was missing from Beacon's installation instructions (or is it implicit in Python -still learning new stuff), otherwise I got an error for missing flask_negotiate.

```
(py3.7) Mikael@mikaellmbp:cgbeacon2$ beacon update genes
Traceback (most recent call last):
  File "/miniconda3/envs/py3.7/bin/beacon", line 33, in <module>
    sys.exit(load_entry_point('cgbeacon2', 'console_scripts', 'beacon')())
  File "/miniconda3/envs/py3.7/bin/beacon", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/miniconda3/envs/py3.7/lib/python3.7/site-packages/importlib_metadata/__init__.py", line 105, in load
    module = import_module(match.group('module'))
  File "/miniconda3/envs/py3.7/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/Mikael/Work/beacon/cgbeacon2/cgbeacon2/cli/__init__.py", line 2, in <module>
    from .commands import cli
  File "/Users/Mikael/Work/beacon/cgbeacon2/cgbeacon2/cli/commands.py", line 7, in <module>
    from cgbeacon2.server import create_app
  File "/Users/Mikael/Work/beacon/cgbeacon2/cgbeacon2/server/__init__.py", line 7, in <module>
    from .blueprints import api_v1
  File "/Users/Mikael/Work/beacon/cgbeacon2/cgbeacon2/server/blueprints/api_v1/__init__.py", line 1, in <module>
    from .views import api1_bp
  File "/Users/Mikael/Work/beacon/cgbeacon2/cgbeacon2/server/blueprints/api_v1/views.py", line 11, in <module>
    from flask_negotiate import consumes
ModuleNotFoundError: No module named 'flask_negotiate'
```

ps. I'll probably not finish this review today.

**Comment by northwestwitch** [comment]
> `pip install -r requirements.txt` was missing from Beacon's installation instructions

Yes, sorry, I forgot to include it in the test instructions! 🙇

**Comment by mikaell** [comment]
I ran into trouble on step #3.

I have edited `scout/server/config.py` to contain

``` 
# -*- coding: utf-8 -*-
BEACON_URL = "http://localhost:6000/apiv1.0"
BEACON_TOKEN = token_created_for_the_beacon_user
```

But when trying to share I get this when clicking 'Add variants to beacon'.

image


See this in the logs:
`2021-01-21 16:10:23 mikaellmbp werkzeug[25183] INFO 127.0.0.1 - - [21/Jan/2021 16:10:23] "POST /beacon_submit HTTP/1.1" 302 -
`

**Comment by northwestwitch** [comment]
> Have you replaced token_created_for_the_beacon_user with the real token created at the step before?

**Comment by mikaell** [comment]
Now Scout's part of the deal seems to be working. But Beacon complains and returns an error.

image


```
INFO:cgbeacon2.utils.auth:Authorized user with id "scout" submits a POST request
ERROR:cgbeacon2.utils.parse:Error while creating VCF iterator from variant file:bad path: scout/demo/643594.clinical.vcf.gz
INFO:werkzeug:127.0.0.1 - - [22/Jan/2021 08:36:23] "POST /apiv1.0/add HTTP/1.1" 422 -
INFO:cgbeacon2.utils.auth:Authorized user with id "scout" submits a POST request
INFO:cgbeacon2.utils.parse:Extracting 263 intervals from the 680 total entries of the VCF file.
ERROR:cgbeacon2.utils.parse:Error while creating VCF iterator from variant file:"intersectBed" does not appear to be installed or on the path, so this method is disabled.  Please install a more recent version of BEDTools and re-import to use this method.
ERROR:cgbeacon2.server:Exception on /apiv1.0/add [POST]
Traceback (most recent call last):
  File "/miniconda3/envs/py3.7/lib/python3.7/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/miniconda3/envs/py3.7/lib/python3.7/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/miniconda3/envs/py3.7/lib/python3.7/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/miniconda3/envs/py3.7/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/miniconda3/envs/py3.7/lib/python3.7/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/miniconda3/envs/py3.7/lib/python3.7/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/Mikael/Work/beacon/cgbeacon2/cgbeacon2/server/blueprints/api_v1/views.py", line 142, in add
    Thread(target=add_variants_task(request)).start()
  File "/Users/Mikael/Work/beacon/cgbeacon2/cgbeacon2/server/blueprints/api_v1/controllers.py", line 85, in add_variants_task
    nr_variants = count_variants(vcf_obj)
  File "/Users/Mikael/Work/beacon/cgbeacon2/cgbeacon2/utils/parse.py", line 230, in count_variants
    for vcf_variant in vcf_obj:
TypeError: 'NoneType' object is not iterable
INFO:werkzeug:127.0.0.1 - - [22/Jan/2021 15:29:19] "POST /apiv1.0/add HTTP/1.1" 500 -
```

**Comment by northwestwitch** [comment]
> TypeError: 'NoneType' object is not iterable

Look like the path to the VCF file is not valid. Have you modified the case in the database providing the full path on your disk to whichever SNV VCF file you want to send to beacon?

For instance if you want to submit both SNV and research SNV files you need to modify both files here:

image

**Comment by mikaell** [comment]
I've updated the config. Both vcf_snv and vcf_snv_research have an absolute path.

-Did you notice the crash is in Beacon's log?

**Comment by mikaell** [comment]
image

**Comment by northwestwitch** [comment]
> I've updated the config. Both vcf_snv and vcf_snv_research have an absolute path.
> 
> -Did you notice the crash is in Beacon's log?

OK! Yes the VCF object is not created not because it doesn't reach destination, but perhaps because of the version of intersectBed. I'll look into it on Monday. Thanks for being the beta tester! Have a nice weekend!

**Comment by northwestwitch** [comment]
@mikaell the reason of the latest error should be caused by a missing bedtools package in the conda environment where the beacon is installed. Bedtools is not a python package and I forgot to mention that you need to install it separately in order to have pybedtools to work. Sorry, it's one of those packages that i have always installed and after a while I forget they don't come with the computer 😄 

Anyway, to install it on your beacon environment:
`conda install -n your_beacon_environment  -c bioconda bedtools`

I'll also include this step in the install instructions if the beacon repo

**Comment by mikaell** [comment]
Yes, after installing BedTools it seems to work! 

image

File: scout/server/blueprints/cases/templates/cases/utils.html [hunk 1]
```diff
@@ -1,5 +1,109 @@
...
+          {% endif %}
+      </div>
+      <div class="modal-footer d-flex justify-content-around">
+        <div>
+          {% if case.beacon %}
+            <a href="{{url_for('cases.beacon_remove', case_id=case._id)}}" class="btn btn-danger text-white">Remove all case variants from Beacon</a>
+          {% else %}
+            <button type="submit" class="btn btn-primary">Add variants to beacon</button>
+          {% endif %}
+          <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
```
Comment (mikaell) [inline scout/server/blueprints/cases/templates/cases/utils.html:98]: I prefere 'Cancel' over 'Close' -but that's more a matter of personal taste.

File: scout/server/blueprints/cases/controllers.py [hunk 2]
```diff
@@ -817,6 +820,131 @@ def multiqc(store, institute_id, case_name):
...
+        flash(
+            f"Variants from the following files are going to be saved to Beacon:{submission['vcf_files']}",
+            "success",
+        )
+        store.case_collection.find_one_and_update(
+            {"_id": case_obj["_id"]}, {"$set": {"beacon": submission}}
+        )
+    return
+
+
```
Comment (mikaell) [inline scout/server/blueprints/cases/controllers.py:947]: This source file is getting really large. Should be refactored some day.

**Review by mikaell — APPROVED** [review]
Looks good! I made two suggestions you can consider.
1. returning empty list explicitly
2. 'Cancel' instead of 'Close'

File: scout/server/blueprints/cases/controllers.py (hunk 2 above)
Comment (northwestwitch) [reply scout/server/blueprints/cases/controllers.py:947]: Yes!

File: scout/server/blueprints/cases/templates/cases/utils.html (hunk 1 above)
Comment (northwestwitch) [reply scout/server/blueprints/cases/templates/cases/utils.html:98]: Cancel sound better, as cancel the whole process and not just close the window, I'll change it!

**Comment by northwestwitch** [comment]
> Looks good! I made two suggestions you can consider.
> 
> 1. returning empty list explicitly
> 2. 'Cancel' instead of 'Close'

I've fixed number 2, number one you fixed it yourself ;)
Thanks for the brave review @mikaell, it was an heavy one!

**Pull request merged**